### PR TITLE
Rmi v3

### PR DESCRIPTION
**Pull request by SamanthaDawsonBanks**
I think this is a the bases of a RMI implementation, it needs a lot of work and some of the methods are just stubs. Additionally there are pages of rem'ed out code that needs cleaning but take a lok see if the logic is sound

### CODE REVIEW

File: mcomp_Java/.settings/org.eclipse.jdt.core.prefs [hunk 1]
```diff
@@ -1,5 +1,6 @@
 eclipse.preferences.version=1
 org.eclipse.jdt.core.compiler.codegen.inlineJsrBytecode=enabled
+org.eclipse.jdt.core.compiler.codegen.methodParameters=do not generate
```
Comment (SamanthaDawsonBanks) [inline mcomp_Java/.settings/org.eclipse.jdt.core.prefs:3]: No idea if this is needed?

**Comment by SamanthaDawsonBanks** [comment]
oh and i want to invert the interfaces (so they all imple remMem/remLdr instead of aggregation)

File: mcomp_Java/.settings/org.eclipse.jdt.core.prefs (hunk 1 above)
Comment (stevempope) [reply mcomp_Java/.settings/org.eclipse.jdt.core.prefs:3]: That should be needed if it’s not a final version so I would say keep for now

File: mcomp_Java/src/common/interfaces/Bossable.java [hunk 2]
```diff
@@ -18,8 +19,8 @@
  */
 public interface Bossable {
 
-  public Map processMapLump(Map m);
+  public Map processMapLump(Herd h);//herd is the data holder for members, maps, and paths
```
Comment (stevempope) [inline mcomp_Java/src/common/interfaces/Bossable.java:22]: This is right I think a lot of our interfaces were scoped right but the methods I attempted as stubs. We may need a meeting to confirm the method signs for our interfaces.

File: mcomp_Java/src/common/interfaces/Connectable.java [hunk 3]
```diff
@@ -34,7 +35,8 @@
    * @param leavingMember The member of the herd that is leaving and this unregistering with the
    *        leader.
    * @return Returns true if the member unregistered with the leader, False if it did not
+   * @throws RemoteException 
    */
-  public boolean deregister(Member leavingMember);
+  public boolean deregister(RemoteMember leavingMember) throws RemoteException;
```
Comment (stevempope) [inline mcomp_Java/src/common/interfaces/Connectable.java:40]: When we talked about tests did we settle on Booleans as the return type of choice? I thought we settled on objects or exceptions if failed?

File: mcomp_Java/src/common/interfaces/Promotable.java [hunk 4]
```diff
@@ -17,6 +18,6 @@
 
 public interface Promotable {
 
-  public Leader becomeLeader(); // TODO returns the initialised Leader object to the Herd?
+  public Leader becomeLeader(Herd h); // TODO returns the initialised Leader object to the Herd?
```
Comment (stevempope) [inline mcomp_Java/src/common/interfaces/Promotable.java:21]: Perhaps this should be a book and the last action on the new leader side is to call a method that says to the members “hi I am the new leader”?

File: mcomp_Java/src/common/objects/Herd.java [hunk 5]
```diff
@@ -125,20 +146,15 @@ public Member nominateLeader() {
...
-              // herdDestSetter.leaveHerd(this);
-            }
+          if (herdDestSetter == null) { // first wins
+            // if (herdDestSetter.getAbilities().size() > 1) {
+            // herdDestSetter.getAbilities().remove(Ability.DEST_SETTER);
+            // } else {
+            // // No longer a leave method
+            // // herdDestSetter.leaveHerd(this);
+            // }
+            herdDestSetter = aspiringMember;
```
Comment (stevempope) [inline mcomp_Java/src/common/objects/Herd.java:158]: AspiringMember?

**Review by stevempope — APPROVED** [review]
This all looks good Dave and must have taken you a long time. It’s really good work. On the initial I think the commented out code looks good too :)

File: mcomp_Java/src/common/objects/Herd.java (hunk 5 above)
Comment (SamanthaDawsonBanks) [reply mcomp_Java/src/common/objects/Herd.java:158]: aspiringMember should be the var of the one requesting to join, as this is in the catch it should be that the first member that is a dest setter gets it and from then on others are not set as it?

**Comment by SamanthaDawsonBanks** [comment]
also some / most / all of the commented out stuff will probably be removed in the long term

**Review by RyanShoobert — APPROVED** [review]

**Pull request merged**